File: lw9/test/Image.test.ts
import { Image } from '../src/Image'; // Путь к вашему файлу с классом Image
import { Tile } from '../src/Tile'; // Путь к вашему файлу с классом Tile
import {beforeEach, describe, expect, jest, test} from "@jest/globals";

describe('Tile Class', () => {
    beforeEach(() => {
        // Сбрасываем счетчик экземпляров перед каждым тестом
        Tile.resetInstanceCount();
    });

    test('should create a tile with default color', () => {
        const tile = new Tile();
        expect(tile.getPixel({ x: 0, y: 0 })).toBe(' '); // Проверяем, что цвет по умолчанию - пробел
    });

    test('should set and get pixel color in tile', () => {
        const tile = new Tile();
        tile.setPixel({ x: 0, y: 0 }, '#');
        expect(tile.getPixel({ x: 0, y: 0 })).toBe('#');
    });

    test('should return space for out of bounds pixel', () => {
        const tile = new Tile();
        expect(tile.getPixel({ x: 8, y: 8 })).toBe(' '); // Проверяем выход за пределы
    });

    test('should count instances of tiles', () => {
        expect(Tile.getInstanceCount()).toBe(0); // Сначала должно быть 0 экземпляров
        const tile1 = new Tile();
        expect(Tile.getInstanceCount()).toBe(1); // Один экземпляр создан
        const tile2 = new Tile();
        expect(Tile.getInstanceCount()).toBe(2); // Два экземпляра созданы
        tile1.setPixel({ x: 0, y: 0 }, '#');
        expect(Tile.getInstanceCount()).toBe(2); // Количество экземпляров не должно измениться
    });
});

describe('Image Class', () => {
    let image: Image;

    beforeEach(() => {
        image = new Image({ width: 16, height: 16 }, '.'); // Создаем новое изображение перед каждым тестом
    });

    test('should get pixel color', () => {
        expect(image.getPixel({ x: 0, y: 0 })).toBe('.'); // Проверяем цвет по умолчанию
    });

    test('should set and get pixel color', () => {
        image.setPixel({ x: 1, y: 1 }, '#');
        expect(image.getPixel({ x: 1, y: 1 })).toBe('#'); // Проверяем установленный цвет
    });

    test('should return space for out of bounds pixel', () => {
        expect(image.getPixel({ x: -1, y: -1 })).toBe(' '); // Проверяем выход за пределы
        expect(image.getPixel({ x: 20, y: 20 })).toBe(' '); // Проверяем выход за пределы
    });

    test('should manage tiles correctly when setting pixels', () => {
        const initialTileCount = Tile.getInstanceCount();

        image.setPixel({ x: 1, y: 1 }, '#');

        expect(Tile.getInstanceCount()).toBe(initialTileCount + 1); // Новый тайл должен быть создан

        image.setPixel({ x: 1, y: 1 }, '@'); // Изменяем цвет в уже существующем тайле

        expect(image.getPixel({ x: 1, y: 1 })).toBe('@'); // Проверяем новый цвет
    });
});

File: lw9_c/test/ImageTest.cpp
#include "catch2/catch_test_macros.hpp"
#include "image.h"

TEST_CASE( "Tile instances count is tracked correctly", "[tile]" ) {
	REQUIRE( Tile::GetInstanceCount() == 0 );
	{
		Tile t1;
		REQUIRE( Tile::GetInstanceCount() == 1 );

		Tile t2('x');
		REQUIRE( Tile::GetInstanceCount() == 2 );
	}
	REQUIRE( Tile::GetInstanceCount() == 0 );
}

TEST_CASE( "Tile operations work correctly", "[tile]" ) {
	Tile t;
	REQUIRE( t.GetPixel({0, 0}) == ' ' );
	REQUIRE( t.GetPixel({7, 7}) == ' ' );
	REQUIRE( t.GetPixel({8, 8}) == ' ' );
	t.SetPixel({0, 0}, 'x');
	REQUIRE( t.GetPixel({0, 0}) == 'x' );
	t.SetPixel({8, 8}, 'y');
	REQUIRE( t.GetPixel({8, 8}) == ' ' );
}

TEST_CASE( "Image instances have correct size", "[image]" ) {
	Image img({8, 8});
	REQUIRE( img.GetSize().width == 8 );
	REQUIRE( img.GetSize().height == 8 );
}

TEST_CASE( "Image pixel operations work correctly", "[image]" ) {
	Image img({8, 8});
	REQUIRE( img.GetPixel({0, 0}) == ' ' );
	REQUIRE( img.GetPixel({7, 7}) == ' ' );
	REQUIRE( img.GetPixel({8, 8}) == ' ' );
	img.SetPixel({0, 0}, 'x');
	REQUIRE( img.GetPixel({0, 0}) == 'x' );
	img.SetPixel({8, 8}, 'y');
	REQUIRE( img.GetPixel({8, 8}) == ' ' );
}

TEST_CASE( "Image LoadImage function works correctly", "[image]" ) {
	std::string pixels = "****\n****\n";
	auto img = LoadImage(pixels);
	REQUIRE( img.GetSize().width == 4 );
	REQUIRE( img.GetSize().height == 2 );

	REQUIRE( img.GetPixel({0, 0}) == '*' );
	REQUIRE( img.GetPixel({3, 1}) == '*' );
	REQUIRE( img.GetPixel({4, 2}) == ' ' );
}

TEST_CASE("Copy on Write works correctly", "[cow][image]") {
	Image img1({8, 8}, 'x');
	Image img2 = img1;

	REQUIRE(img1.GetSize().width == img2.GetSize().width);
	REQUIRE(img1.GetSize().height == img2.GetSize().height);

	img1.SetPixel({0, 0}, 'y');
	REQUIRE(img1.GetPixel({0, 0}) != img2.GetPixel({0, 0}));
}

TEST_CASE("Copy on Write correctly duplicates tiles", "[cow][tile][image]") {
	Image img({32, 32}, 'x');

	int initialTileCount = Tile::GetInstanceCount();

	Image img2 = img;

	REQUIRE(Tile::GetInstanceCount() == initialTileCount);

	img.SetPixel({0, 0}, 'y');

	REQUIRE(Tile::GetInstanceCount() == initialTileCount + 1);
}